### PR TITLE
Signup: Add progress state handling in Redux

### PR DESCRIPTION
**Pull request by jsnmoon**
Part of step 2 for #6709 (following advice found [here](https://github.com/Automattic/wp-calypso/pull/24490#discussion_r186011164)). 

Builds off of #25606 which enables reducers to consume steps and flows data without DOM dependency. 

The intent of this PR is to eventually handle all of the signup processes through Redux instead of Flux. Specifically, this PR adds Redux handling (e.g. actions, reducers, and selectors) for signup progress handling. A separate PR will refactor SignupProgressStore to dispatch Redux actions instead of Flux actions.

Note that code introduced in this PR doesn't interact with any part of our application (yet).

### CODE REVIEW

**Comment by jsnmoon** [comment]
@Automattic/team-calypso, any thoughts on this change? This is part of reduxifying the signup stores. 

It necessitates merging #25606 before it can land, however.

**Comment by blowery** [comment]
Taking a look!

**Comment by blowery** [comment]
Looks good to me! Took me a minute to realize it wasn't targeting master, but #25606. @jsnajdr any thoughts?

**Comment by jsnmoon** [comment]
I'll change the parent to `master` once #25606 lands :)

File: client/state/signup/progress/reducer.js [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+		return updateStep( state, { ...step, status } );
+	}
+
+	return addStep( state, { ...step, status } );
+}
+
+//
+// Helper Functions
+//
+export function addStep( state, step ) {
```
Comment (jsnajdr) [inline client/state/signup/progress/reducer.js]: This export (and also `updateStep` and `processStep`) exists only for tests, doesn't it? What if the module exported only one thing, the reducer? That's the only export that's really needed by the app. And it's a unit that's small enough to be unit tested as a whole. `addStep` already gets a complete test coverage with the actions that use it. There's no need to test it separately.

I find it hard to figure out modules that export almost everything. When reading the code, I don't know what's the public API of the module.

Besides this little nit, the code looks pretty solid.

What is your plan to handle Redux state persistence for signup? There are two special behaviors that are tricky to implement with our Redux persistence helpers:
- the persisted progress store should be ignored if the current flow name doesn't match the persisted one (see calls to `fetchCachedSignup`)
- the signup flow is the only Calypso feature that crosses the logged-out/logged-in boundary. The store persisted in logged-out state won't be loaded in the logged-in state.

Comment (jsnmoon) [reply client/state/signup/progress/reducer.js]: > This export (and also `updateStep` and `processStep`) exists only for tests, doesn't it? (@jsnajdr)

Correct. However, `processStep` actually shouldn't be exported; I'll fix that.

> What if the module exported only one thing, the reducer? That's the only export that's really needed by the app. And it's a unit that's small enough to be unit tested as a whole. addStep already gets a complete test coverage with the actions that use it. There's no need to test it separately. (@jsnajdr)

Good point. I originally architected the reducer with the assumption that most action handlers would essentially be a call to `updateStep`. However, that ended up not being the case.

I'll look into eliminating excess exports.

> What is your plan to handle Redux state persistence for signup? There are two special behaviors that are tricky to implement with our Redux persistence helpers:  (@jsnajdr)

> (1) the signup flow is the only Calypso feature that crosses the logged-out/logged-in boundary. The store persisted in logged-out state won't be loaded in the logged-in state.

I'm working off the assumption that the progress store (and the dependency store) is only useful to the logged-out user while they are progressing through the signup flow. Once they complete the flow by creating an account, discarding the logged-out state sounds like the right way to go. 

> (2) There are two special behaviors that are tricky to implement with our Redux persistence helpers: the persisted progress store should be ignored if the current flow name doesn't match the persisted one (see calls to `fetchCachedSignup`) (@jsnajdr)

@taggon recently introduced [a function](https://github.com/Automattic/wp-calypso/commit/aa21a468d1c17aa6267448c8368fcc815cc784b7) to discard steps not associated with the current flow. 

I intend to invoke this function in the constructor of the [`signup flow controller's constructor (where fetchCachedSignup is being invoked)`](https://github.com/Automattic/wp-calypso/blob/4df0c252c0e43aa98749ac5fdd20ff7503198db0/client/lib/signup/flow-controller.js#L63) to clear out any progress not associated with the current flow. Alternatively, we can indiscriminately clear out the progress state when initializing a new signup flow controller.

Thoughts?

**Review by jsnajdr — APPROVED** [review]
Thanks for improving the tests and the module's public API! 🥇

**Comment by jsnmoon** [comment]
Now that #25606 has landed, I rebased this branch off of master. 

Smoke testing locally on my end just in case before landing this in the next hour.

**Pull request merged**